### DISCUSSION TITLE
comcom64 works

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#2151] [General]
Finally! After so many years. :)
All the stuff is already in PPA.
But the comcom64 integration
in dosemu2 is not enabled in PPA
because there is no dj64 build
for arm yet. So if someone wants
to try that out, dosemu2 should
be built by hands.

Boot process works fine, commands
seems to work. But its likely still a
bit unstable. In fact, its started for
the first time today, so definitely
there is a lot of debugging ahead.
Any help, as  always, welcome!

Now if someone wonders why it
takes the whole ~345Kb, then just do
`djstrip comcom64`
and it is reduced to ~180Kb, while
comcom32 is ~195Kb stripped!
So comcom64 is smaller than comcom32
even now. In the future it will probably
be much smaller than today.
gdb debugging is fully supported,
and that was quite a challenge to
implement. If you happen to rebuild
comcom64 locally, make sure to have
the latest djstub package from today,
to get the reliable gdb support.

There are still 2 major things to do
(in addition to the routine testing/debugging
of course):

- arm port.
I unfortunately failed with an idea of
having a platform-independent build.
Yes, its a bit weird to have the DOS
program under dosemu separately
built for x86 and arm, but for now this
is all I can offer.

- get rid of gcc-ia16 requirement in djstub...
Unless I am mistaken, gcc-ia16 is not in
any major distro, so, similar to djgpp, it
represents the problem for promoting
dosemu2 to such distros. But unlike with
djgpp, which is used to build the entire
comcom32, here gcc-ia16 is only used to
build djstub. But unfortunately djstub
itself have grown to the source code size
comparable with that of comcom32 itself,
so I am risking to fall into an infinite recursion
rewriting tool-chains. :)

PS:
Aleksey Navalny, RIP!
This comcom64 release is dedicated to him.

**Comment by stsp** [comment]
I switched the PPA build to comcom64.
Please tell me if you observe something
unusual with that change.

**Reply by stsp** [reply]
I also disabled the comcom32 daily
builds and disconnected the djgpp repo.
Unless some comcom64 bugs pops up,
its probably the time to say goodbye to
comcom32 and djgpp. :)

For now I use the pre-built stub to
avoid gcc-ia16. In the future, some
solution to build the stub from sources
again, should be developed.

**Comment by stsp** [comment]
> its probably the time to say goodbye to
> comcom32

Except that I now added a 32bit build
to comcom64's CI, and it builds comcom32
as a test! That's really cool, as it shows how
we can build both versions from the same
source tree, meaning that the dj64 tool-chain
is very compatible. This is unlike fdpp,
where the sources were modified quite
enough to make the real-mode build a real
challenge.

So... comcom32 is not really dead.
It silently exists as a build possibility
of comcom64. Just no deployments.

**Reply by andrewbird** [reply]
I'm adding a Github Actions build under DOS for FreeDOS kernel repo using Dosemu2/FDPP. It was working a few weeks ago, but I now have an unmet dependency.
~~~
 The following packages have unmet dependencies:
 dosemu2 : Depends: dj64 but it is not installable
           Depends: comcom64 but it is not going to be installed
~~~
What PPA am I missing / shouldn't this be part of Dosemu2 PPA?

**Reply by stsp** [reply]
https://code.launchpad.net/~stsp-0/+archive/ubuntu/dj64
This one.

**Reply by andrewbird** [reply]
Thanks, that fixed the missing dependency.

**Comment by stsp** [comment]
> dosemu2 : Depends: dj64 but it is not installable

Hmm, does this mean that the
install is now broken for all users?
I missed that fact that adding a ppa
deps on LP doesn't auto-add such
ppa as a dependency...

**Reply by andrewbird** [reply]
Not sure but I just ran this on my 22.04 ubuntu (I've been waiting until the new comcom64 stuff settled down)
~~~
root@calypso:~# apt dist-upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
Get more security updates through Ubuntu Pro with 'esm-apps' enabled:
  libvlc5 exo-utils vlc-data libvlccore9 imagemagick libjs-jquery-ui
  libopenexr25 python3-scipy libpostproc55 libmagickcore-6.q16-6-extra
  libavcodec58 libmagickwand-6.q16-6 libavutil56 imagemagick-6.q16 libswscale5
  libmagickcore-6.q16-6 libexo-2-0 libgsl27 libswresample3
  imagemagick-6-common vlc-plugin-video-output libavformat58 libgslcblas0
  libvlc-bin vlc-plugin-base libexo-common libavfilter7
Learn more about Ubuntu Pro at https://ubuntu.com/pro
The following packages have been kept back:
  dosemu2 firmware-sof-signed
The following packages will be upgraded:
  fdpp gigolo tcpdump thunk-gen
4 to upgrade, 0 to newly install, 0 to remove and 2 not to upgrade.
Need to get 1,033 kB of archives.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] n
Abort.
~~~

Can't you move the dj64 over to Dosemu2 PPA?

**Reply by stsp** [reply]
So I don't see any error in the example
you pasted.
But I added a second build recipe to
build into dosemu2 ppa.
Strange that people didn't report any
dependency problems. But I noted an
increased visiting count of my dj64 page
on github, with the referrer being google,
so obviously people were fighting this
on their own. :(

**Comment by andrewbird** [comment]
The problem was this
~~~
The following packages have been kept back:
  dosemu2 firmware-sof-signed
~~~

**Reply by stsp** [reply]
Ahha!
Good.
So users were not affected mostly.
They just were not getting an update.
This is probably what @WayCon2 was
saying in #2160 but I didn't understand.

So now users will finally get the updates -
exactly when comcom64 have stabilized.
Phew, very good. :)

**Reply by andrewbird** [reply]
> But I noted an
increased visiting count of my dj64 page
on github, with the referrer being google,

I think your comcom64 got a mention on freedos-devel, so perhaps that fueled it.

**Reply by stsp** [reply]
Was it mentioned such vaguely that
people would use google? :) I mean,
github shows the referrer, so I see
where people are coming from - not
from the freedos list.
Btw, today I posted an announce to
djgpp ML out of nothing to do. :)
I don't think there are dosemu2 users
there, but who knows.

**Reply by andrewbird** [reply]
I think people may have visited comcom64's github and since there is no readme looked at the commit log and thought 'dj64' what's that.

**Reply by stsp** [reply]
> I think your comcom64 got a mention on freedos-devel, so perhaps that fueled it.

Hmm, it appears also comcom32 was
mentioned. With @lproven asking
"why should anyone care?"
The answer is: no one should.
We in dosemu2 no longer care about
comcom32, but if someone wants the
32bit command.com in freedos, then
he can write a readme on his own.
I believe that at least comcom32 eats
much fewer conventional memory than
any other command.coms. Being a
prot-mode app, it doesn't need the conv
memory. Maybe there are some other
bonuses, but definitely not much.

**Reply by stsp** [reply]
> I think people may have visited comcom64's github and since there is no readme looked at the commit log
> and thought 'dj64' what's that.

Maybe, although if you are already at
my github page, why not to look among
my repos first. :)
OK so you don't support the guess that
people could fight a dependency problem.
And in fact it is good if they didn't.

**Reply by andrewbird** [reply]
> OK so you don't support the guess that
people could fight a dependency problem.

I just think they likely didn't notice there was one, unless like me (on Github Actions) they were doing a first time install of Dosemu2.

**Reply by stsp** [reply]
> I think people may have visited comcom64's github and since there is no readme

But do we really need a readme for
command.com? Does it have to explain
what command.com is?

**Reply by andrewbird** [reply]
I don't think it has to explain what command.com is, but perhaps to say what advantages if any it has over FreeCOM, or why you use it on Dosemu/FDPP.

BTW the updates came now to my local machine and I was also able to drop the stsp-0/dj64 repo from my FreeDOS kernel/PR, thanks!

**Reply by stsp** [reply]
> but perhaps to say what advantages if any it has over FreeCOM

It probably doesn't have many.

> or why you use it on Dosemu/FDPP.

Because it can be built by a 64bit
GPLed tool-chain. I don't think this
point is convincing for many people.
Of course in the future it may became
better than freecom... if it absorbs as
much development efforts as fdpp got
to out-run freedos. But its not an intention.

**Reply by stsp** [reply]
README.md is here:
https://github.com/dosemu2/comcom64/blob/master/README.md
Note the new URL, its now in dosemu2
space. In fact, I simply renamed the
comcom32 repository. :) Since they have
the common history, no need for a new
repo.

**Comment by stsp** [comment]
Hmm, either no one is using dosemu2
these days, or comcom64 is bug-free?
No single bug report?
Quite alarming. :)

**Reply by jschwartzenberg** [reply]
I haven't checked it out just yet, but I'm definitely planning to!!

**Comment by j2969719** [comment]
I would like to figure out what dependencies I need these days to build dosemu2-git+fdpp-git+comcom64 under archlinux?
from what I understood and saw in aur, to build fdpp I will need not just nasm, but https://github.com/stsp/nasm-segelf and https://github.com/stsp/thunk_gen
and for comcom64: https://github.com/stsp/dj64dev  (https://github.com/stsp/thunk_gen) and https://github.com/stsp/djstub ?

Also in aur a dosemu package is created along with some freedos binaries, their presence in the package is not critical if I keep such stuff in the drive_c folder?

**Reply by stsp** [reply]
This is correct.
You can look up all the deps in debian/control.
To build djstub, you need lfanew+smallerc.
So yes, quite a large dependency chain.

> Also in aur a dosemu package is created along with some freedos binaries,
> their presence in the package is not critical if I keep such stuff in the drive_c folder?

I don't know which ones, but yes,
nothing from freedos is critical.
We provide install-freedos scripts
for those who need freedos tools.

**Reply by stsp** [reply]
This is correct.
You can look up all the deps in debian/control.
To build djstub, you need lfanew+smallerc.
So yes, quite a large dependency chain.

> Also in aur a dosemu package is created along with some freedos binaries,
> their presence in the package is not critical if I keep such stuff in the drive_c folder?

I don't know which ones, but yes,
nothing from freedos is critical.
We provide install-freedos scripts
for those who need freedos tools.

**Reply by stsp** [reply]
Note that comcom32 is still buildable
with `make 32`. You can still use it.
Given that dosemu2 is that difficult to
build these days, I decided to not remove
deprecated things.

**Reply by stsp** [reply]
Also note some quirks: for fdpp you
need clang (not gcc) for compiling, but
binutils (not llvm's lld or gold) for linking.
For stripping, only llvm-strip may work
(binutils's strip already have the needed
patch, but its probably not yet in arch).
So really one have to cherry-pick the tools
from 2 tool-chains... :(

**Reply by j2969719** [reply]
tnx
btw, should the comcom32 and comcom64 packages conflict, or is it possible to install both at the same time and switch between them via some kind of userhook or something?

**Reply by stsp** [reply]
Switching is not anticipated.
Just comcom64 takes the preference.
If its not there, comcom32 is searched.
So yes, you can install both, but comcom32
will then be useless.

**Reply by j2969719** [reply]
Got it, thanks a lot for your help.

**Comment by stsp** [comment]
I am glad to announce that comcom64
(and therefore also comcom32 if someone
builds it) finally got all the missing features.
That includes ae0x interfacing and tab
completion.
"Extended" features, like mouse support
and internal MEM command, are not yet.
But they are not so important.

So with host-gdb debugging, that was
just a routine task.
Does anyone want to finally implement
MEM, or the mouse support? :)

**Comment by stsp** [comment]
Mouse control is now implemented.
You can type with mouse by clicking
the chars on screen, and you can
tab-complete with right button, as
well as do many other things.
I updated README.md to document
the mouse controls.

Basically you don't need keyboard
any more. :) I wonder what other
command.com has such an advanced
mouse control (if at all)?

**Reply by Torinde** [reply]
Good. Can you also hook in the horizontal scroll (tilt wheel) function? Back/Forward mouse buttons as well, but those are less common.

You may want to check [DOSbox-X mouse wheel support](https://dosbox-x.com/wiki/#Guide:Mouse-support-in-DOSBox‐X#_mouse_wheel_support) - it's less focused on command line functionality, but maybe you'll like some of the features.

**Reply by stsp** [reply]
Good point.
2 problems here:
- I follow the ctmouse wheel API: https://cutemouse.sourceforge.net/wheelapi.txt
- I don't have a mouse with horizontal scroll wheel just yet

So to get this work done, you need to
extend the wheel API and preferably
talk to ctmouse devs about your
extensions, making it official. And I
will need to buy the new mouse.

Or if dosbox-X has its own int33 extensions
for wheel, I'd be happy to implement
those. But I don't want to shoehorn my
own extensions in that area - there are
already so many dosemu2-specific
extensions, that you can't, for example,
run comcom64 under real DOS.

**Reply by stsp** [reply]
You may also want to talk to
vbados devs https://git.javispedro.com/cgit/vbados.git/about/
about the int33 h-wheel API,
as they are also a recognized
players in that area. I don't care
whom to stay compatible with,
but lets be compatible at least
with something.

**Reply by Torinde** [reply]
It seems Logitech already made such INT33 extensions for Cyberman! And there is source code available for interfacing with them:
- https://github.com/joncampbell123/dosbox-x/issues/1271

I made a post about this discussion in the above Issue.

**Comment by stsp** [comment]
Today I've found a very interesting
"bug" in comcom64, which I initially
fixed... before realizing what it actually
was! As it was already fixed, I didn't
want to revert the fix, so instead I
added the `mouseopt` command with
a few options. `/C` option enables the
aforementioned "bug", which actually
allows to control other software with
mouse.

For example. You can now do:
```
mouseopt /c
freecom
```
This will run freecom, which is mouse-unaware
by itself. But as the result of a "bug"
(which is now a feature) you can control
it with mouse absolutely similar to comcom64!
I.e. scroll the history buffer with the wheel,
type by clicking letters on screen, tab-complete
with RMB, and so on. It just became fully
mouse-aware!
And yes, this discovery was entirely occasional. :)
I had to apply quite a lot of patches to
dosemu2 so that this bug not to crash
win31, but now it works reliably.

**Comment by stsp** [comment]
Btw, this comcom64 is not something new.
Its actually just another closed regression.
Not many of you know that dosemu-1.2
already had the 64bit command.com
(called comcom) on the coopthreads
infrastructure. It was implemented
somewhere around year 2001
(27a24db6 54afa890) by Hans Lermen
in a couple of months.
In contrast to this, I started re-implementing
coopthreads within dosemu2 infrastructure
in 2012 (736c2628) and finished the DPMI
integration in 2021 (5c6cf0d6), so 9 years.
Then 3 more years went to port comcom32
to that infrastructure. I didn't even write
comcom32, I took it from the leaked Centroid
sources, whereas in dosemu1 comcom was
written from zero, and was 64bit from the
beginning.
So why it took 12 years to just re-implement
what was written in a few months, is a bit
unclear, and this is not even the last regression
here. We still don't have a replacement for
dosemu1's dexe format for distributing programs.
And given the initial failure of probonopd to
even create a working appimage of dosemu2,
I fear this will take up the next decade... :)

**Reply by stsp** [reply]
> and was 64bit from the
> beginning.

Ah, except that dosemu-1.2 had no
64bit ports, so it was actually all 32bit...
Well, then comcom64 is at least
somewhat new. :)

**Comment by stsp** [comment]
Hi guys!

I've finally modularized the dj64dev
suite, so the amount of dj64 packages
increased twice.
If anyone knows how to create the build
recipes on LP such that the recipe builds
only a subset of the packages instead of
all, please let me know, as so far I haven't
found such possibility.

For Arch packaging hopefully no extra
work is needed, i.e. the monolithic build
should still work as before. Let me know
if it broke, and optionally you can follow
the split if you want.

I also updated the README:
https://github.com/stsp/dj64dev
It should now be clear what is dj64dev,
djdev64, dj64-devel, dj64dev-djdev64-devel
and all other combinations of such. :)
I also outlined the DPMI requirements
for supporting dj64 programs, so whoever
writes DPMI servers here, may be interested
to take a look. Unfortunately I don't expect
it to be easy to integrate the dj64 support
into hdpmi, as IIRC hdpmi is asm-written,
and for dj64 the C-written DPMI server
suits much better. But I don't know if it
ever existed in a wild.

**Reply by j2969719** [reply]
Is djstub now part of dj64dev or does it still need to be compiled from a separate djstub repository?

**Reply by stsp** [reply]
Its separate.
It needs another compiler (smallerc),
so adding it to dj64dev is not a good
idea (like adding binutils into gcc).

**Reply by stsp** [reply]
BUT: please note that lfanew
package is no longer needed.

**Comment by j2969719** [comment]
judging by `ERROR: SIGILL while in vm86...`it looks like I really screwed up with my crappy PKGBUILDs somewhere(
welp I’d better use it from aur since it’s already there

**Reply by stsp** [reply]
Try comcom32.
The chances are good that something
in the comcom64's stack was not
properly re-built for you.

**Reply by j2969719** [reply]
It seems like at some point I installed smallerc from aur.  in my old smalerc PKGBUILD I simply used the source code from your fork with strip disabled, after rebuilding with it the problem went away for some reason ¯\_(ツ)_/¯ Is it safe to use your fork or is an upstream like in aur necessary now?

**Reply by stsp** [reply]
There are no changes between my fork
and upstream, except for the debian
build rules.
